File: Models/BiomaSurfacePartonSoilSWATC/src/stics/BiomaSurfacePartonSoilSWATC/SurfacePartonSoilSWATCComponent.f90
MODULE Surfacepartonsoilswatcmod
    USE Surfacetemperaturepartonmod
    USE Soiltemperatureswatmod
    IMPLICIT NONE
CONTAINS

    SUBROUTINE model_surfacepartonsoilswatc(DayLength, &
        GlobalSolarRadiation, &
        AboveGroundBiomass, &
        AirTemperatureMinimum, &
        AirTemperatureMaximum, &
        LayerThickness, &
        BulkDensity, &
        SoilProfileDepth, &
        AirTemperatureAnnualAverage, &
        VolumetricWaterContent, &
        LagCoefficient, &
        SurfaceTemperatureMinimum, &
        SurfaceTemperatureMaximum, &
        SurfaceSoilTemperature, &
        SoilTemperatureByLayers)
        IMPLICIT NONE
        INTEGER:: i_cyml_r
        REAL, INTENT(IN) :: DayLength
        REAL, INTENT(IN) :: GlobalSolarRadiation
        REAL, INTENT(IN) :: AboveGroundBiomass
        REAL, INTENT(IN) :: AirTemperatureMinimum
        REAL, INTENT(IN) :: AirTemperatureMaximum
        REAL , DIMENSION(: ), INTENT(IN) :: LayerThickness
        REAL , DIMENSION(: ), INTENT(IN) :: BulkDensity
        REAL, INTENT(IN) :: SoilProfileDepth
        REAL, INTENT(IN) :: AirTemperatureAnnualAverage
        REAL , DIMENSION(: ), INTENT(IN) :: VolumetricWaterContent
        REAL, INTENT(IN) :: LagCoefficient
        REAL, INTENT(OUT) :: SurfaceTemperatureMinimum
        REAL, INTENT(OUT) :: SurfaceTemperatureMaximum
        REAL, INTENT(OUT) :: SurfaceSoilTemperature
        REAL , DIMENSION(: ), INTENT(OUT) :: SoilTemperatureByLayers
        !- Name: SurfacePartonSoilSWATC -Version: 001, -Time step: 1
        !- Description:
    !            * Title: SurfacePartonSoilSWATC model
    !            * Authors: simone.bregaglio@unimi.it
    !            * Reference: ('http://bioma.jrc.ec.europa.eu/ontology/JRC_MARS_biophysical_domain.owl',)
    !            * Institution: Universiy Of Milan
    !            * ExtendedDescription: Composite strategy for the calculation of surface temperature with Parton's method and soil temperature with SWAT method. Parton, W. J. 1984. Predicting soil temperatures in a shortgrass steppe. Soil Science 138:93-101. and Neitsch,S.L., Arnold, J.G., Kiniry, J.R., Williams, J.R., King, K.W. Soil and Water Assessment Tool. Theoretical documentation. Version 2000. http://swatmodel.tamu.edu/media/1290/swat2000theory.pdf.  Composite strategy. See also references of the associated strategies.
    !            * ShortDescription: None
        !- inputs:
    !            * name: DayLength
    !                          ** description : Length of the day
    !                          ** inputtype : variable
    !                          ** variablecategory : exogenous
    !                          ** datatype : DOUBLE
    !                          ** max : 24
    !                          ** min : 0
    !                          ** default : 10
    !                          ** unit : h
    !            * name: GlobalSolarRadiation
    !                          ** description : Daily global solar radiation
    !                          ** inputtype : variable
    !                          ** variablecategory : exogenous
    !                          ** datatype : DOUBLE
    !                          ** max : 50
    !                          ** min : 0
    !                          ** default : 15
    !                          ** unit : Mj m-2 d-1
    !            * name: AboveGroundBiomass
    !                          ** description : Above ground biomass
    !                          ** inputtype : variable
    !                          ** variablecategory : exogenous
    !                          ** datatype : DOUBLE
    !                          ** max : 60
    !                          ** min : 0
    !                          ** default : 3
    !                          ** unit : Kg ha-1
    !            * name: AirTemperatureMinimum
    !                          ** description : Minimum daily air temperature
    !                          ** inputtype : variable
    !                          ** variablecategory : exogenous
    !                          ** datatype : DOUBLE
    !                          ** max : 50
    !                          ** min : -60
    !                          ** default : 5
    !                          ** unit : 
    !            * name: AirTemperatureMaximum
    !                          ** description : Maximum daily air temperature
    !                          ** inputtype : variable
    !                          ** variablecategory : exogenous
    !                          ** datatype : DOUBLE
    !                          ** max : 60
    !                          ** min : -40
    !                          ** default : 15
    !                          ** unit : 
    !            * name: LayerThickness
    !                          ** description : Soil layer thickness
    !                          ** inputtype : parameter
    !                          ** parametercategory : constant
    !                          ** datatype : DOUBLEARRAY
    !                          ** len : 
    !                          ** max : 3
    !                          ** min : 0.005
    !                          ** default : 0.05
    !                          ** unit : m
    !            * name: BulkDensity
    !                          ** description : Bulk density
    !                          ** inputtype : parameter
    !                          ** parametercategory : constant
    !                          ** datatype : DOUBLEARRAY
    !                          ** len : 
    !                          ** max : 1.8
    !                          ** min : 0.9
    !                          ** default : 1.3
    !                          ** unit : t m-3
    !            * name: SoilProfileDepth
    !                          ** description : Soil profile depth
    !                          ** inputtype : parameter
    !                          ** parametercategory : constant
    !                          ** datatype : DOUBLE
    !                          ** max : 50
    !                          ** min : 0
    !                          ** default : 3
    !                          ** unit : m
    !            * name: AirTemperatureAnnualAverage
    !                          ** description : Annual average air temperature
    !                          ** inputtype : parameter
    !                          ** parametercategory : constant
    !                          ** datatype : DOUBLE
    !                          ** max : 50
    !                          ** min : -40
    !                          ** default : 15
    !                          ** unit : degC
    !            * name: VolumetricWaterContent
    !                          ** description : Volumetric soil water content
    !                          ** inputtype : variable
    !                          ** variablecategory : exogenous
    !                          ** datatype : DOUBLEARRAY
    !                          ** len : 
    !                          ** max : 0.8
    !                          ** min : 0
    !                          ** default : 0.25
    !                          ** unit : m3 m-3
    !            * name: LagCoefficient
    !                          ** description : Lag coefficient that controls the influence of the previous day's temperature on the current day's temperature
    !                          ** inputtype : parameter
    !                          ** parametercategory : constant
    !                          ** datatype : DOUBLE
    !                          ** max : 1
    !                          ** min : 0
    !                          ** default : 0.8
    !                          ** unit : dimensionless
        !- outputs:
    !            * name: SurfaceTemperatureMinimum
    !                          ** description : Minimum surface soil temperature
    !                          ** datatype : DOUBLE
    !                          ** variablecategory : auxiliary
    !                          ** max : 60
    !                          ** min : -60
    !                          ** unit : degC
    !            * name: SurfaceTemperatureMaximum
    !                          ** description : Maximum surface soil temperature
    !                          ** datatype : DOUBLE
    !                          ** variablecategory : auxiliary
    !                          ** max : 60
    !                          ** min : -60
    !                          ** unit : degC
    !            * name: SurfaceSoilTemperature
    !                          ** description : Average surface soil temperature
    !                          ** datatype : DOUBLE
    !                          ** variablecategory : auxiliary
    !                          ** max : 60
    !                          ** min : -60
    !                          ** unit : degC
    !            * name: SoilTemperatureByLayers
    !                          ** description : Soil temperature of each layer
    !                          ** datatype : DOUBLEARRAY
    !                          ** variablecategory : state
    !                          ** len : 
    !                          ** max : 60
    !                          ** min : -60
    !                          ** unit : degC
        call model_surfacetemperatureparton(DayLength, AirTemperatureMaximum,  &
                AirTemperatureMinimum, AboveGroundBiomass,  &
                GlobalSolarRadiation,SurfaceTemperatureMinimum,SurfaceTemperatureMaximum,SurfaceSoilTemperature)
        call model_soiltemperatureswat(VolumetricWaterContent,  &
                SurfaceSoilTemperature, LayerThickness, LagCoefficient,  &
                SoilTemperatureByLayers, AirTemperatureAnnualAverage, BulkDensity,  &
                SoilProfileDepth)
    END SUBROUTINE model_surfacepartonsoilswatc

END MODULE

File: Models/BiomaSurfacePartonSoilSWATC/src/apsim/BiomaSurfacePartonSoilSWATC/SurfacePartonSoilSWATCComponent.cs
using Models.Core;
using Models.Utilities;
using System; 
namespace Models.Crop2ML;
     

/// <summary>
///  SurfacePartonSoilSWATC component
/// </summary>
public class SurfacePartonSoilSWATCComponent 
{

    /// <summary>
    ///  constructor of SurfacePartonSoilSWATC component
    /// </summary>
    public SurfacePartonSoilSWATCComponent() {}

    //Declaration of the associated strategies
    SurfaceTemperatureParton _SurfaceTemperatureParton = new SurfaceTemperatureParton();
    SoilTemperatureSWAT _SoilTemperatureSWAT = new SoilTemperatureSWAT();

    /// <summary>
    /// Gets and sets the Soil layer thickness
    /// </summary>
    [Description("Soil layer thickness")] 
    [Units("m")] 
    public double[] LayerThickness
    {
        get
        {
             return _SoilTemperatureSWAT.LayerThickness; 
        }
        set
        {
            _SoilTemperatureSWAT.LayerThickness = value;
        }
    }

    /// <summary>
    /// Gets and sets the Bulk density
    /// </summary>
    [Description("Bulk density")] 
    [Units("t m-3")] 
    public double[] BulkDensity
    {
        get
        {
             return _SoilTemperatureSWAT.BulkDensity; 
        }
        set
        {
            _SoilTemperatureSWAT.BulkDensity = value;
        }
    }

    /// <summary>
    /// Gets and sets the Soil profile depth
    /// </summary>
    [Description("Soil profile depth")] 
    [Units("m")] 
    public double SoilProfileDepth
    {
        get
        {
             return _SoilTemperatureSWAT.SoilProfileDepth; 
        }
        set
        {
            _SoilTemperatureSWAT.SoilProfileDepth = value;
        }
    }

    /// <summary>
    /// Gets and sets the Annual average air temperature
    /// </summary>
    [Description("Annual average air temperature")] 
    [Units("degC")] 
    public double AirTemperatureAnnualAverage
    {
        get
        {
             return _SoilTemperatureSWAT.AirTemperatureAnnualAverage; 
        }
        set
        {
            _SoilTemperatureSWAT.AirTemperatureAnnualAverage = value;
        }
    }

    /// <summary>
    /// Gets and sets the Lag coefficient that controls the influence of the previous day's temperature on the current day's temperature
    /// </summary>
    [Description("Lag coefficient that controls the influence of the previous day's temperature on the current day's temperature")] 
    [Units("dimensionless")] 
    public double LagCoefficient
    {
        get
        {
             return _SoilTemperatureSWAT.LagCoefficient; 
        }
        set
        {
            _SoilTemperatureSWAT.LagCoefficient = value;
        }
    }

    /// <summary>
    /// Algorithm of SurfacePartonSoilSWATC component
    /// </summary>
    public void CalculateModel(SurfacePartonSoilSWATCState s,SurfacePartonSoilSWATCState s1,SurfacePartonSoilSWATCRate r,SurfacePartonSoilSWATCAuxiliary a,SurfacePartonSoilSWATCExogenous ex)
    {
        _SurfaceTemperatureParton.CalculateModel(s,s1, r, a, ex);
        _SoilTemperatureSWAT.CalculateModel(s,s1, r, a, ex);
    }

    /// <summary>
    /// Initialization of SurfacePartonSoilSWATC component
    /// </summary>
    public void Init(SurfacePartonSoilSWATCState s, SurfacePartonSoilSWATCState s1, SurfacePartonSoilSWATCRate r, SurfacePartonSoilSWATCAuxiliary a, SurfacePartonSoilSWATCExogenous ex)
    {
        _SoilTemperatureSWAT.Init(s, s1, r, a, ex);
    }

    /// <summary>
    /// constructor copy of SurfacePartonSoilSWATC component
    /// </summary>
    /// <param name="toCopy"></param>
    public SurfacePartonSoilSWATCComponent(SurfacePartonSoilSWATCComponent toCopy): this() // copy constructor 
    {
            
            for (int i = 0; i < toCopy.LayerThickness.Length; i++)
                { LayerThickness[i] = toCopy.LayerThickness[i]; }
    
            
            for (int i = 0; i < toCopy.BulkDensity.Length; i++)
                { BulkDensity[i] = toCopy.BulkDensity[i]; }
    
            SoilProfileDepth = toCopy.SoilProfileDepth;
            AirTemperatureAnnualAverage = toCopy.AirTemperatureAnnualAverage;
            LagCoefficient = toCopy.LagCoefficient;
    }
}